File: App/src/Adaptive.ReactiveTrader.Client.Domain/ReactiveTrader.cs
using System;
using System.Reactive.Linq;
using Adaptive.ReactiveTrader.Client.Domain.Models;
using Adaptive.ReactiveTrader.Client.Domain.Repositories;
using Adaptive.ReactiveTrader.Client.Domain.ServiceClients;
using Adaptive.ReactiveTrader.Client.Domain.Transport;
using log4net;

namespace Adaptive.ReactiveTrader.Client.Domain
{
    public class ReactiveTrader : IReactiveTrader
    {
        private ConnectionProvider _connectionProvider;
        private static readonly ILog Log = LogManager.GetLogger(typeof(ReactiveTrader));

        public void Initialize(string username, string[] servers)
        {
            _connectionProvider = new ConnectionProvider(username, servers);

            var referenceDataServiceClient = new ReferenceDataServiceClient(_connectionProvider);
            var executionServiceClient = new ExecutionServiceClient(_connectionProvider);
            var blotterServiceClient = new BlotterServiceClient(_connectionProvider);
            var pricingServiceClient = new PricingServiceClient(_connectionProvider);

            var tradeFactory = new TradeFactory();
            var executionRepository = new ExecutionRepository(executionServiceClient, tradeFactory);
            var priceFactory = new PriceFactory(executionRepository);
            var priceRepository = new PriceRepository(pricingServiceClient, priceFactory);
            var currencyPairFactory = new CurrencyPairFactory(priceRepository);
            TradeRepository = new TradeRepository(blotterServiceClient, tradeFactory);
            ReferenceData = new ReferenceDataRepository(referenceDataServiceClient, currencyPairFactory);
        }

        public IReferenceDataRepository ReferenceData { get; private set; }
        public ITradeRepository TradeRepository { get; private set; }

        public IObservable<ConnectionInfo> ConnectionStatus
        {
            get
            {
                return _connectionProvider.GetActiveConnection()
                    .Do(_ => Log.Info("New connection created by connection provider"))
                    .Select(c => c.Status)
                    .Switch()
                    .Publish()
                    .RefCount();
            }
        }
    }
}

File: App/src/Adaptive.ReactiveTrader.Client.JS/ReactiveTrader.ts
 class ReactiveTrader implements IReactiveTrader {
     
     private _connection: IConnection;

     initialize(username: string, server: string): Rx.Observable<{}> {
         this._connection = new Connection(server, username);

         var referenceDataServiceClient = new ReferenceDataServiceClient(this._connection);
         var executionServiceClient = new ExecutionServiceClient(this._connection);
         var blotterServiceClient = new BlotterServiceClient(this._connection);
         var pricingServiceClient = new PricingServiceClient(this._connection);

         var tradeFactory = new TradeFactory();
         var executionRepository = new ExecutionRepository(executionServiceClient, tradeFactory);
         var priceFactory = new PriceFactory(executionRepository);
         var priceRepository = new PriceRepository(pricingServiceClient, priceFactory);
         var currencyPairUpdateFactory = new CurrencyPairUpdateFactory(priceRepository);

         this.tradeRepository = new TradeRepository(blotterServiceClient, tradeFactory);
         this.referenceDataRepository = new ReferenceDataRepository(referenceDataServiceClient, currencyPairUpdateFactory);

         return this._connection.initialize();
     }

     tradeRepository: ITradeRepository;
     referenceDataRepository: IReferenceDataRepository;

     get connectionStatusStream(): Rx.Observable<ConnectionInfo> {
         return this._connection
             .status
             .publish()
             .refCount();
     }
 }

